<launch>

    <!-- mono or stereo and what ros bag to play -->
    <arg name="max_cameras" default="2" />
    <arg name="use_stereo"  default="true" />
    <arg name="bag_start"   default="20" /> <!-- v1-2: 0, mh1: 40, mh2: 35, mh3-5: 20 -->


    <!-- imu starting thresholds -->
    <arg name="init_window_time"  default="0.75" />   <!-- MH04 0.5 others 0.75-->
    <arg name="init_imu_thresh"   default="1.5" />

    <!-- saving trajectory path and timing information -->
    <arg name="dosave"      default="true" />
    <arg name="dotime"      default="true" />
    <arg name="path_est"    default="/tmp/traj_estimate.txt" />
    <arg name="path_rect_est"    default="/tmp/traj_rect_estimate.txt" />
    <arg name="path_transform"    default="/tmp/transform.txt" />
    <arg name="path_transform_1"   default="/tmp/transform1.txt" />
    <arg name="path_transform_2"   default="/tmp/transform2.txt" />
    <arg name="path_transform_3"   default="/tmp/transform3.txt" />
    <arg name="path_transform_4"   default="/tmp/transform4.txt" />
    <arg name="path_time"   default="/tmp/traj_timing.txt" />
    <arg name="path_local_relative_transform" default="/tmp/local_relative_transform.txt" />
   

    <arg name="use_prior_map"  default="true" />
    <arg name="trans_fej" default="true" />
    <arg name="use_fej" default="true" />
    <arg name="use_schmidt" default="true" />

    <!-- multi map setting -->
    <!-- how many maps to use -->
    <arg name="num_map" default="4" />
    <arg name="pose_graph_filename_1" default="qMH03_rMH01_r2d2_loc_timeq_timer_2dq_2dr_3dr.txt" />
    <arg name="keyframe_pose_filename_1" default="traj_estimate_MH01_cam.txt" />
    <arg name="pose_graph_filename_2" default="qMH03_rMH02_r2d2_loc_timeq_timer_2dq_2dr_3dr.txt" />
    <arg name="keyframe_pose_filename_2" default="traj_estimate_MH02_cam.txt" />
    <arg name="pose_graph_filename_3" default="qMH03_rMH04_r2d2_loc_timeq_timer_2dq_2dr_3dr.txt" />
    <arg name="keyframe_pose_filename_3" default="traj_estimate_MH04_cam.txt" />
    <arg name="pose_graph_filename_4" default="qMH03_rMH05_r2d2_loc_timeq_timer_2dq_2dr_3dr.txt" />
    <arg name="keyframe_pose_filename_4" default="traj_estimate_MH05_cam.txt" />


    <!-- MASTER NODE! -->
    <node name="run_subscribe_msckf" pkg="ov_msckf" type="run_subscribe_msckf" output="screen" clear_params="true" required="true">
    <!-- <node name="run_subscribe_msckf" pkg="ov_msckf" type="run_subscribe_msckf" output="screen" clear_params="true" required="true" launch-prefix="gdb -ex run &#45;&#45;args"> -->
        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/imu0" />
        <param name="topic_camera0"  type="string" value="/cam0/image_raw" />
        <param name="topic_camera1"  type="string" value="/cam1/image_raw" />
        
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V1_01_easy.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V1_02_medium.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V1_03_difficult.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V2_01_easy.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V2_02_medium.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/V2_03_difficult.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_01_easy.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_02_easy.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_03_medium.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_04_difficult.txt" /> -->
        <!-- <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH_05_difficult.txt" /> -->
        <param name="path_gt"     type="string" value="$(find ov_data)/euroc_mav/MH03_aligned.txt" />



        <!-- match map parameters-->
       <param name="use_prior_map" type="bool" value="$(arg use_prior_map)" />
        <param name="trans_fej" type="bool" value="$(arg trans_fej)" />
        <param name="use_schmidt" type="bool" value="$(arg use_schmidt)" />
        <param name="map_save_path" type="string" value="/home/zzq/Code/C-MIMB-VIO/src/C-MIMB-VIO/matches/" /> 
        
        <param name="num_map" type="int" value="$(arg num_map)" />
        <param name="pose_graph_filename_1" type="string" value="$(arg pose_graph_filename_1)" />
        <param name="keyframe_pose_filename_1" type="string" value="$(arg keyframe_pose_filename_1)" />
        <param name="pose_graph_filename_2" type="string" value="$(arg pose_graph_filename_2)" />
        <param name="keyframe_pose_filename_2" type="string" value="$(arg keyframe_pose_filename_2)" />
        <param name="pose_graph_filename_3" type="string" value="$(arg pose_graph_filename_3)" />
        <param name="keyframe_pose_filename_3" type="string" value="$(arg keyframe_pose_filename_3)" />
        <param name="pose_graph_filename_4" type="string" value="$(arg pose_graph_filename_4)" />
        <param name="keyframe_pose_filename_4" type="string" value="$(arg keyframe_pose_filename_4)" /> 
        <param name="load_map_traj" type="bool" value="true" />
        

        <param name="points_filename" type="string" value="/tmp/match_points.txt" />
        <param name="save_match_points" type="bool" value="true" />
        <param name="match_num_thred" type="int" value="30" />
        <param name="pixel_error" type="double" value="3.0" />
        <param name="transform_cov" type="double" value="10.0" />
        <param name="keyframe_cov_pos" type="double" value="0.0001"/>
        <param name="keyframe_cov_ori" type="double" value="0.00025"/>

        <param name="iter_num" type="int" value="0"/>
        <param name="opt_thred" type="double" value="1000"/>

        <param name="kf_match_interval" type="double" value="0" />
        <param name="use_robust_match" type="bool" value="true" />
        <param name="show_match_image" type="bool" value="false" /> 

        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="$(arg use_fej)" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="$(arg use_stereo)" />
        <param name="calib_cam_extrinsics"   type="bool"   value="false" />
        <param name="calib_cam_intrinsics"   type="bool"   value="false" />
        <param name="calib_cam_timeoffset"   type="bool"   value="false" /> <!--false-->
        <param name="calib_camimu_dt"        type="double" value="0.0" />
        <param name="max_clones"             type="int"    value="11" />
        <param name="max_slam"               type="int"    value="25" /> <!-- 25-->
        <param name="max_slam_in_update"     type="int"    value="25" /> <!-- 25 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="999" />
        <param name="max_cameras"            type="int"    value="$(arg max_cameras)" />
        <param name="dt_slam_delay"          type="double" value="3" />
        <param name="init_window_time"       type="double" value="$(arg init_window_time)" />
        <param name="init_imu_thresh"        type="double" value="$(arg init_imu_thresh)" />
        <rosparam param="gravity">[0.0,0.0,9.81]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- zero velocity update parameters -->
        <param name="try_zupt"               type="bool"   value="false" />
        <param name="zupt_chi2_multipler"    type="int"    value="2" />
        <param name="zupt_max_velocity"      type="double" value="0.3" />
        <param name="zupt_noise_multiplier"  type="double" value="50" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="$(arg dotime)" />
        <param name="record_timing_filepath"      type="string" value="$(arg path_time)" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"            type="bool"   value="true" />
        <param name="num_pts"            type="int"    value="150" />
        <param name="fast_threshold"     type="int"    value="15" />
        <param name="grid_x"             type="int"    value="5" />
        <param name="grid_y"             type="int"    value="3" />
        <param name="min_px_dist"        type="int"    value="10" />
        <param name="knn_ratio"          type="double" value="0.70" />
        <param name="downsample_cameras" type="bool"   value="false" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="1.6968e-04" />
        <param name="gyroscope_random_walk"        type="double"   value="1.9393e-05" />
        <param name="accelerometer_noise_density"  type="double"   value="2.0000e-3" />
        <param name="accelerometer_random_walk"    type="double"   value="3.0000e-3" />


        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[752, 480]</rosparam>   <!-- rosparam以yaml来读数据，通常用于矩阵的读取-->
        <rosparam param="cam1_wh">[752, 480]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="false" />
        <param name="cam1_is_fisheye" type="bool" value="false" />
        <rosparam param="cam0_k">[458.654,457.296,367.215,248.375]</rosparam>
        <rosparam param="cam0_d">[-0.28340811,0.07395907,0.00019359,1.76187114e-05]</rosparam>
        <rosparam param="cam1_k">[457.587,456.134,379.999,255.238]</rosparam>
        <rosparam param="cam1_d">[-0.28368365,0.07451284,-0.00010473,-3.55590700e-05]</rosparam>

        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
            0.0148655429818, -0.999880929698, 0.00414029679422, -0.0216401454975,
            0.999557249008, 0.0149672133247, 0.025715529948, -0.064676986768,
            -0.0257744366974, 0.00375618835797, 0.999660727178, 0.00981073058949,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>
        <rosparam param="T_C1toI">
            [
            0.0125552670891, -0.999755099723, 0.0182237714554, -0.0198435579556,
            0.999598781151, 0.0130119051815, 0.0251588363115, 0.0453689425024,
            -0.0253898008918, 0.0179005838253, 0.999517347078, 0.00786212447038,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>


    </node>
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find ov_msckf)/launch/display_2.rviz" />

    <!-- play the dataset -->
    <!-- <node pkg="rosbag" type="play" name="rosbag" args="-d 5 -r 1 -s $(arg bag_start) $(arg bag)" required="true"/> -->

    <!-- record the trajectory if enabled -->
    <group if="$(arg dosave)">
        <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen" required="true">
            <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
        </node>
         <node name="recorder_estimate_rect" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/poseimu_rect" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_rect_est)" />

        </node>
        <node name="recorder_transform" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/transform" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_transform)" />
        </node>
         <node name="recorder_local_relative_transform" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/local_relative_transform" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_local_relative_transform)" />
        </node>
        <node name="recorder_transform_1" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/transform1" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_transform_1)" />
        </node>
        <node name="recorder_transform_2" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/transform2" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_transform_2)" />
        </node>
        <node name="recorder_transform_3" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/transform3" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_transform_3)" />
        </node>
        <node name="recorder_transform_4" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/transform4" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_transform_4)" />
        </node>
    </group>


</launch>